// p-s/dist/get-logger.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _console = require('console');

var _console2 = _interopRequireDefault(_console);

var _arrify = require('arrify');

var _arrify2 = _interopRequireDefault(_arrify);

var _lodash = require('lodash');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var _require = require('../package.json'),
    version = _require.version;

var shouldLog = {
  error: getShouldLogFn('error'),
  warn: getShouldLogFn('warn', 'error'),
  info: getShouldLogFn('info', 'warn', 'error')
};

exports.default = getLogger;


function getLogger(logLevel) {
  return {
    error: getLogFn('error'),
    warn: getLogFn('warn'),
    info: getLogFn('info')
  };

  function getLogFn(name) {
    return function logFn() {
      if (shouldLog[name](logLevel || process.env.LOG_LEVEL)) {
        var message = getMessage.apply(undefined, arguments);
        _console2.default[name].apply(_console2.default, _toConsumableArray(message)); // eslint-disable-line no-console
      }
    };
  }
}

function getMessage(first) {
  for (var _len = arguments.length, rest = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    rest[_key - 1] = arguments[_key];
  }

  if ((0, _lodash.isPlainObject)(first) && first.message && first.ref) {
    return [].concat(_toConsumableArray((0, _arrify2.default)(first.message)), [getLink(first.ref), first.error], rest).filter(function (i) {
      return !!i;
    });
  } else {
    return [first].concat(rest);
  }
}

function getLink(ref) {
  return 'https://github.com/kentcdodds/p-s/blob/v' + version + '/other/ERRORS_AND_WARNINGS.md#' + ref;
}

function getShouldLogFn() {
  for (var _len2 = arguments.length, acceptableValues = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
    acceptableValues[_key2] = arguments[_key2];
  }

  acceptableValues = ['', 'debug'].concat(_toConsumableArray(acceptableValues));
  return function shouldLogWithLevel() {
    var logLevel = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';

    logLevel = logLevel.toLowerCase();
    return !logLevel || acceptableValues.some(function (v) {
      return logLevel === v;
    });
  };
}
